x/streamer/keeper/distribute.go

```diff
@@ -0,0 +1,174 @@
+		return err
+	}
+	return nil
+}
+
+// moveActiveStreamToFinishedStream moves a stream that has completed its distribution from an active to a finished status.
+func (k Keeper) moveActiveStreamToFinishedStream(ctx sdk.Context, stream types.Stream) error {
+	timeKey := getTimeKey(stream.StartTime)
+	if err := k.deleteStreamRefByKey(ctx, combineKeys(types.KeyPrefixActiveStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	if err := k.addStreamRefByKey(ctx, combineKeys(types.KeyPrefixFinishedStreams, timeKey), stream.Id); err != nil {
+		return err
+	}
+	return nil
+}
+
+// distributeInternal runs the distribution logic for a stream, and adds the sends to
+// the distrInfo struct. It also updates the stream for the distribution.
+func (k Keeper) distributeInternal(
+	ctx sdk.Context, stream types.Stream,
+) (sdk.Coins, error) {
+	totalDistrCoins := sdk.NewCoins()
+	remainCoins := stream.Coins.Sub(stream.DistributedCoins...)
+	remainEpochs := uint64(stream.NumEpochsPaidOver - stream.FilledEpochs)
+
+	for _, coin := range remainCoins {
+		epochAmt := coin.Amount.Quo(sdk.NewInt(int64(remainEpochs)))
+		if epochAmt.IsPositive() {
+			newlyDistributedCoin := sdk.Coin{Denom: coin.Denom, Amount: epochAmt}
+			totalDistrCoins = totalDistrCoins.Add(newlyDistributedCoin)
+		}
+	}
+	totalDistrCoins = totalDistrCoins.Sort()
+
+	err := k.updateStreamPostDistribute(ctx, stream, totalDistrCoins)
+	if err != nil {
+		return nil, err
+	}
+
+	distAddr, err := sdk.AccAddressFromBech32(stream.DistributeTo)
+	if err != nil {
+		return nil, err
+	}
+
+	err = k.bk.SendCoinsFromModuleToAccount(
+		ctx,
+		types.ModuleName,
+		distAddr,
+		totalDistrCoins,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx.EventManager().EmitEvents(sdk.Events{
+		sdk.NewEvent(
+			types.TypeEvtDistribution,
+			sdk.NewAttribute(types.AttributeReceiver, distAddr.String()),
+			sdk.NewAttribute(types.AttributeAmount, totalDistrCoins.String()),
+		),
+	})
+	return totalDistrCoins, nil
+}
+
+// updateStreamPostDistribute increments the stream's filled epochs field.
+// Also adds the coins that were just distributed to the stream's distributed coins field.
+func (k Keeper) updateStreamPostDistribute(ctx sdk.Context, stream types.Stream, newlyDistributedCoins sdk.Coins) error {
+	stream.FilledEpochs += 1
+	stream.DistributedCoins = stream.DistributedCoins.Add(newlyDistributedCoins...)
+	if err := k.setStream(ctx, &stream); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Distribute distributes coins from an array of streams to all eligible locks.
+func (k Keeper) Distribute(ctx sdk.Context, streams []types.Stream) (sdk.Coins, error) {
+	totalDistributedCoins := sdk.Coins{}
+	for _, stream := range streams {
+		streamDistributedCoins, err := k.distributeInternal(ctx, stream)
+		if err != nil {
+			return nil, err
+		}
+		totalDistributedCoins = totalDistributedCoins.Add(streamDistributedCoins...)
+	}
+
+	k.checkFinishDistribution(ctx, streams)
```

Comment:
why iterate again and not just check it in the same loop vs doing 2 loops? 
probably in the `updateStreamPostDistribute` the checkDistribution can be done and then no need for the comparison with the +1 in the epochs (line 154).